wifiphisher/common/extensions.py

```diff
@@ -0,0 +1,294 @@
+                            continue
+                    else:
+                        break
+
+    def set_interface(self, interface):
+        """
+        Sets interface for EM.
+
+        :param interface: Interface name
+        :type interface: String
+        :return: None
+        :rtype: None
+        """
+
+        self._interface = interface
+        self._socket = linux.L2Socket(iface=self._interface)
+
+    def set_extensions(self, extensions):
+        """
+        Sets extensions for EM.
+
+        :param extensions: List of str extension names
+        :type extensions: List
+        :return: None
+        :rtype: None
+        """
+
+        self._extensions_str = extensions
+
+    def init_extensions(self, shared_data):
+        """
+        Init EM extensions. Should be run
+        when all shared data has been gathered.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :param shared_data: Dictionary object
+        :type shared_data: Dictionary
+        :return: None
+        :rtype: None
+        """
+
+        # Convert shared_data from dict to named tuple
+        shared_data = collections.namedtuple('GenericDict',
+                                             shared_data.keys())(**shared_data)
+        # Initialize all extensions with the shared data
+        for extension in self._extensions_str:
+            mod = importlib.import_module(
+                constants.EXTENSIONS_LOADPATH + extension)
+            ExtensionClass = getattr(mod, extension.title())
+            obj = ExtensionClass(shared_data)
+            self._extensions.append(obj)
+
+    def start_extensions(self):
+        """
+        Starts the two main daemons of EM:
+
+        1) Daemon that listens to every packet and
+        forwards it to each extension for further processing.
+        2) Daemon that receives special-crafted packets
+        from extensions and broadcasts them in the air.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :return: None
+        :rtype: None
+        """
+
+        # One daemon is listening for packets...
+        self._listen_thread.start()
+        # ...another daemon is sending packets
+        self._send_thread.start()
+        # daemon for channel hopping
+        self.get_channels()
+        self._channelhop_thread.start()
+
+    def on_exit(self):
+        """
+        Stops both daemons of EM on exit.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :return: None
+        :rtype: None
+        """
+
+        self._should_continue = False
+        if self._listen_thread.is_alive():
+            self._listen_thread.join(3)
+        if self._send_thread.is_alive():
+            self._send_thread.join(3)
+        if self._channelhop_thread.is_alive():
+            self._send_thread.join(3)
+        self._socket.close()
+
```

Comment:
`self._socket.close` will cause problem when `-nJ` option is used. Since it is not being initialized when there is no jamming.